go/internal/feast/sqliteonlinestore.go

```diff
@@ -0,0 +1,147 @@
+	"google.golang.org/protobuf/proto"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type SqliteOnlineStore struct {
+	// Feast project name
+	project string
+	path    string
+	db      *sql.DB
+	db_mu   sync.Mutex
+}
+
+// Creates a new sqlite online store object. onlineStoreConfig should have relative path of database file with respect to repoConfig.repoPath.
+func NewSqliteOnlineStore(project string, repoConfig *RepoConfig, onlineStoreConfig map[string]interface{}) (*SqliteOnlineStore, error) {
+	store := SqliteOnlineStore{project: project}
+	if db_path, ok := onlineStoreConfig["path"]; !ok {
+		return nil, fmt.Errorf("cannot find sqlite path %s", db_path)
+	} else {
+		if dbPathStr, ok := db_path.(string); !ok {
+			return nil, fmt.Errorf("cannot find convert sqlite path to string %s", db_path)
+		} else {
+			store.path = fmt.Sprintf("%s/%s", repoConfig.RepoPath, dbPathStr)
+			db, err := initializeConnection(store.path)
+			if err != nil {
+				return nil, err
+			}
+			store.db = db
+		}
+	}
+
+	return &store, nil
+}
+
+func (s *SqliteOnlineStore) Destruct() {
+	s.db.Close()
+}
+
+// Returns FeatureData 2D array. Each row corresponds to one entity value and each column corresponds to a single feature where the number of columns should be
+// same length as the length of featureNames. Reads from every table in featureViewNames with the entity keys described.
+func (s *SqliteOnlineStore) OnlineRead(ctx context.Context, entityKeys []types.EntityKey, featureViewNames []string, featureNames []string) ([][]FeatureData, error) {
+	featureCount := len(featureNames)
+	_, err := s.getConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	project := s.project
+	results := make([][]FeatureData, len(entityKeys))
+	entityNameToEntityIndex := make(map[string]int)
+	in_query := make([]string, len(entityKeys))
+	serialized_entities := make([]interface{}, len(entityKeys))
+	for i := 0; i < len(entityKeys); i++ {
+		serKey, err := serializeEntityKey(&entityKeys[i])
+		if err != nil {
+			return nil, err
+		}
+		// TODO: fix this, string conversion is not safe
+		entityNameToEntityIndex[string(*serKey)] = i
+		// for IN clause in read query
+		in_query[i] = "?"
+		serialized_entities[i] = *serKey
+	}
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	for idx := 0; idx < len(entityKeys); idx++ {
+		results[idx] = make([]FeatureData, featureCount)
+	}
+	for _, featureViewName := range featureViewNames {
+		query_string := fmt.Sprintf(`SELECT entity_key, feature_name, value, event_ts
+									FROM %s
+									WHERE entity_key IN (%s)
+									ORDER BY entity_key`, tableId(project, featureViewName), strings.Join(in_query, ","))
+		rows, err := s.db.Query(query_string, serialized_entities...)
+		if err != nil {
+			return nil, err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var entity_key string
+			var feature_name string
+			var valueString string
```

Comment:
can we read bytes from sqlite? seems like we can loose data here as well

Reply:
Changed to byte []